core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,268 @@
+            } else {
+                processTrailers((HttpHeaders) data, outputQueue::addLast);
+            }
+            return;
+        }
+
+        if (data instanceof HttpData) {
+            input.add(convertToByteBuf((HttpData) data));
+            process(input, outputQueue::addLast);
+            input.discardReadBytes();
+        }
+    }
+
+    /**
+     * Decodes an informational {@link ResponseHeaders} to N objects.
+     */
+    protected void processInformationalHeaders(ResponseHeaders in, HttpDeframerOutput<T> out) {}
+
+    /**
+     * Decodes an non-informational {@link HttpHeaders} to N objects.
+     */
+    protected void processHeaders(HttpHeaders in, HttpDeframerOutput<T> out) {}
+
+    /**
+     * Decodes a {@link HttpHeaders trailers} to N objects.
+     */
+    protected void processTrailers(HttpHeaders in, HttpDeframerOutput<T> out) {}
+
+    /**
+     * Invoked when a {@link Throwable} is raised while deframing.
+     */
+    protected void processOnError(Throwable cause) {}
+
+    @Override
+    public final void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstreamUpdater.compareAndSet(this, null, subscription)) {
+            deferredInit();
+        } else {
+            subscription.cancel();
+        }
+    }
+
+    @Override
+    final SubscriptionImpl subscribe(SubscriptionImpl subscription) {
+        final SubscriptionImpl subscriptionImpl = super.subscribe(subscription);
+        if (subscribedUpdater.compareAndSet(this, 0, 1)) {
+            deferredInit();
+        }
+        return subscriptionImpl;
+    }
+
+    private void deferredInit() {
+        if (upstream != null && subscribed != 0) {
+            if (initializedUpdater.compareAndSet(this, 0, 1)) {
+                if (demand() > 0) {
+                    upstream.request(1);
+                }
+            }
+        }
+    }
+
+    @Override
+    void request(long n) {
+        if (initialized != 0 && outputQueue.isEmpty()) {
+            upstream.request(1);
+        }
+        super.request(n);
+    }
+
+    @Override
+    void cancel() {
+        cancelAndCleanup();
+        super.cancel();
+    }
+
+    @Override
+    public final void onNext(HttpObject data) {
+        if (eventLoop.inEventLoop()) {
+            onNext0(data);
+        } else {
+            eventLoop.execute(() -> onNext0(data));
+        }
+    }
+
+    private void onNext0(HttpObject data) {
+        try {
+            process(data);
+
+            if (outputQueue.isEmpty()) {
+                upstream.request(1);
+            } else {
+                for (;;) {
+                    final T deframed = outputQueue.poll();
+                    if (deframed != null) {
+                        write(deframed);
+                    } else {
+                        break;
+                    }
+                }
+                whenConsumed().thenRun(() -> {
+                    if (demand() > 0) {
```

Comment:
demand is visible from the downstream subscription thread.
How we guarantee `eventLoop` is the same thread?
Should we use the thread from the downstream `SubscriptionImpl` instead?

Reply:
>  How we guarantee eventLoop is the same thread?
>  Should we use the thread from the downstream SubscriptionImpl instead?

That's a good point. The publisher is subscribed internally and only used the request-scoped thread. That means the `eventLoop` of `HttpDeframer` is the same as downstream's thread.
As a micro-optimization, I think it works safely.

By the way, if you think this assumption is hard-coded, I would fix it to use `SubscriptionImpl`'s thread instead of injecting it from the constructor.

Reply:
If so, how about adding a logic that checks the two threads are the same?